sky/serve/serve_utils.py

```diff
@@ -611,6 +616,116 @@ def _get_replica_status() -> serve_state.ReplicaStatus:
     return ''
 
 
+def extract_replica_id_from_launch_log_file_name(file_name: str) -> int:
+    pattern = r'.*replica_(\d+)_launch.log'
+    match = re.search(pattern, file_name)
+    if match:
+        return int(match.group(1))
+    raise ValueError(f'Failed to get replica id from file name: {file_name}')
+
+
+def has_valid_replica_id(file_name: str,
+                         target_replica_id: Optional[int]) -> bool:
+    if target_replica_id is None:
+        return True
+    replica_id = extract_replica_id_from_launch_log_file_name(file_name)
+    return replica_id == target_replica_id
+
+
+def prepare_replica_logs_for_download(service_name: str, timestamp: str,
+                                      target_replica_id: Optional[int]) -> None:
+    logger.info('Preparing replica logs for download')
+    remote_service_dir_name = generate_remote_service_dir_name(service_name)
+    dir_name = os.path.expanduser(remote_service_dir_name)
+    dir_for_download = os.path.join(dir_name, timestamp)
+    os.makedirs(dir_for_download, exist_ok=True)
+
+    # copy over log files of replicas already terminated
+    log_file_pattern = os.path.join(dir_name, 'replica_*.log')
+    log_files = glob.glob(log_file_pattern)
+    terminated_replica_log_files = [
+        file for file in log_files if not file.endswith('_launch.log') and
+        has_valid_replica_id(file, target_replica_id)
+    ]
+    for file_path in terminated_replica_log_files:
+        shutil.copy(file_path, dir_for_download)
+
+    # manage log files of replicas with launch log files
+    launch_log_files = [
+        file for file in log_files if file.endswith('_launch.log') and
+        has_valid_replica_id(file, target_replica_id)
+    ]
+    for launch_log_file in launch_log_files:
+        replica_id = extract_replica_id_from_launch_log_file_name(
+            launch_log_file)
+        replica_info = serve_state.get_replica_info_from_id(
+            service_name, replica_id)
+        if replica_info is None:
```

Comment:
For replica that has been normally terminated (scale down), we don't keep the record in database but we will keep its log file under the service directory. So it is likely the `replica_info is None` IIUC. Should we just `continue` here?

Reply:
Inside this `for` loop, we are only dealing with launch log files, which shouldn't be relevant to replicas that have been normally terminated. For replicas that have been normally terminated, we manage them here 
https://github.com/skypilot-org/skypilot/blob/1eb3a5301faf219e4b3e0896425fdb191072e376/sky/serve/serve_utils.py#L657-L662